### PR TITLE
Add ArmorTypes to HitShape

### LINKED ISSUE #7718: Add engine support for features required by a MechCommander-based RTS mod
Have always wanted to make one, as I see massive fun-potential in this. 
To my knowledge MechCommander (Gold) is freeware (though not open-source like MechCommander 2, unfortunately), so there shouldn't be any legal issues.

There are several features and fixes required to make this good and keep the amount of work in check (especially on the graphics side). This issue is supposed to keep track of the individual requirements.
- [x] #7717 Make actor height (and Z height for targeting) customisable
- [x] #7716 Add support for SSI shp version 1.10 (assembly outside of OpenRA)
- [ ] #9268 Add support for directional armor
- [x] #7666 Support for using individual files for each frame of a sequence
- [x] #6631 Add SplitFacings option to various Render traits
- [x] #6465 Minor offset issues with barrels, muzzle flashes...
- [x] #6247 Issue with order click area (diamond-shaped cells)
- [x] #3736 Missiles move towards ground before moving towards target
- [x] #9677 Facing reflection/mirroring (required for #3274 as well)
- [x] #7675 Raycast collision detection for projectiles

### LINKED ISSUE #9268: Implement support for directional armor
A possible, relatively simple approach would be to compare the projectile facing right before explosion with the facing of directly hit actors (explosion happened inside their health radius / hit box / whatever).

If the projectile facing exactly matched or diverted by no more than X degrees, it must have hit the weak rear armor, giving a damage bonus.
If the facing was the opposite, it must have hit the front, reducing the damage.
In between it must have hit the sides, so apply normal damage.

This is just a possibility. If we had a raycaster and rectangular hit boxes, it would be possible to check more accurately what side was hit.

### PR DESCRIPTION
**Pull request by reaperrr**
This allows to assign Armor traits to specific HitShapes, allowing to make certain parts of actors more vulnerable or durable than others.
Can also be used to simulate directional armor in a good-enough way that this closes #7718.

Testcase can be found here: https://github.com/reaperrr/OpenRA/tree/hitshape-armor-test

Closes #9268

### CODE REVIEW

**Comment by obrakmann** [comment]
Of course this means that we would have to add the ability to drive in reverse to prevent units from showing the weaker rear armor when retreating.

**Comment by ltem** [comment]
The testcase works great.

-------------------------

Following obrakmann's reverse-drive-ability note, a way to position units like  the Tiberium Wars Formation Move would also be useful.

|image|
|----|

**Comment by pchote** [comment]
I don't think we shoudl add directional armor or reverse move to our default mods, but I see no reason why other mods couldn't do both (aside from Mobile/Move being crap).

File: OpenRA.Mods.Common/Warheads/DamageWarhead.cs [hunk 1]
```diff
@@ -63,5 +74,27 @@ public virtual void DoImpact(Actor victim, Actor firedBy, IEnumerable<int> damag
...
+		}
+
+		static bool ValidArmorType(string armor, HashSet<string> armorTypes)
+		{
+			if (armorTypes.Count == 0)
+				return true;
+
+			// PERF: Avoid LINQ.
+			foreach (var at in armorTypes)
+				if (armor.StartsWith(at))
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Warheads/DamageWarhead.cs]: Why are you using `StartsWith`? This should also get a `StringComparison.Ordinal`.

**Review by abcdefg30 — DISMISSED** [review]
Looks good to me otherwise.

File: OpenRA.Mods.Common/Warheads/DamageWarhead.cs (hunk 1 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Warheads/DamageWarhead.cs]: I wasn't sure how spaces are handled, nor what else to use instead.

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/HitShape.cs [hunk 2]
```diff
@@ -25,6 +25,10 @@ public class HitShapeInfo : ConditionalTraitInfo, Requires<BodyOrientationInfo>
 		[Desc("Create a targetable position at the center of each occupied cell. Stacks with TargetableOffsets.")]
 		public readonly bool UseTargetableCellsOffsets = false;
 
+		[Desc("Defines which Armor types apply when the actor receives damage to this HitShape.",
+			"If none specified, all armor types the actor has are valid.")]
+		public readonly HashSet<string> ArmorTypes = new HashSet<string>();
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/HitShape.cs]: I'll see if I can come up with something, at least for TargetDamageWarhead.

**Comment by reaperrr** [comment]
Updated. I haven't had time to test the Target/HealthPercentageWH changes yet, though.

File: OpenRA.Mods.Common/Warheads/HealthPercentageDamageWarhead.cs [hunk 3]
```diff
@@ -43,12 +43,20 @@ public override void DoImpact(Actor victim, Actor firedBy, IEnumerable<int> dama
...
 			var healthInfo = victim.Info.TraitInfoOrDefault<HealthInfo>();
 			if (healthInfo == null)
 				return;
 
+			// Cannot be damaged without an active HitShape
+			var activeShapes = victim.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled);
+			if (!activeShapes.Any())
+				return;
+
+			var closestActiveShape = activeShapes.MinBy(t => t.Info.Type.DistanceFromEdge(victim.CenterPosition, victim));
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Warheads/HealthPercentageDamageWarhead.cs]: One pass option:
```c#
var closestActiveShape = victim.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled)
    .MinByOrDefault(t => t.Info.Type.DistanceFromEdge(victim.CenterPosition, victim));
if (closestActiveShape == null)
    return;
```
Requires you to allocate the MinBy lambda even if there are no active hit shapes, but I'm assuming that's the uncommon case/a worthwhile tradeoff.

**Comment by reaperrr** [comment]
Updated.

**Comment by reaperrr** [comment]
@OpenRA/engine-hackers This is basically blocking some related follow-ups, so it would be nice if this could get another review soon.

**Review by pchote — COMMENTED** [review]
I really want to be able to :+1: this, but i'm sorry to say that this really is still setting off my future-maintenance-nightmare spidey-sense.

The core of the problem is that this introduces a second channel for applying damage (DoImpact with and without a hitshape) right in the middle of a chain of inheritance and overrides.  This makes it very difficult to know which will be used, and means that the same weapon/warheads/vs fired by different means will end up doing different amounts of damage to the same actor.

This feature really needs to be integrated from the bottom up, and not just bolted on the side.  Require *all* damaging warheads to specify the hitshape, and fix the higher level code to be able to choose the appropriate one based on whatever rules make sense in each case.

**Comment by pchote** [comment]
There isn't a clear way forwards here, so moving to the Future milestone.

**Pull request closed without merging**